kitsune/wiki/views.py

```diff
@@ -452,125 +452,154 @@ def steal_lock(request, document_slug, revision_id=None):
 
+    if doc.locale != request.LANGUAGE_CODE:
+        # We've fallen back to the parent, since no visible translation existed.
+        url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
+        return HttpResponseRedirect(url)
+
     # If this document has a parent, then the edit is handled by the
     # translate view. Pass it on.
     if doc.parent:
         return translate(request, doc.parent.slug, revision_id)
-    if revision_id:
-        rev = get_object_or_404(Revision, pk=revision_id, document=doc)
-    else:
-        rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]
 
-    disclose_description = bool(request.GET.get("opendescription"))
-    doc_form = rev_form = None
-    if doc.allows(user, "create_revision"):
-        rev_form = RevisionForm(instance=rev, initial={"based_on": rev.id, "comment": ""})
-    if doc.allows(user, "edit"):
-        doc_form = DocumentForm(
-            initial=_document_form_initial(doc),
-            can_archive=can_archive,
-            can_edit_needs_change=can_edit_needs_change,
-        )
-
-    if request.method == "GET":
-        if not (rev_form or doc_form):
-            # You can't do anything on this page, so get lost.
-            raise PermissionDenied
-
-    else:  # POST
-        # Comparing against localized names for the Save button bothers me, so
-        # I embedded a hidden input:
-        which_form = request.POST.get("form")
+    rev = _get_rev(doc, revision_id)
+    doc_form = DocumentForm(
+        initial=_document_form_initial(doc),
+        can_archive=can_archive,
+        can_edit_needs_change=can_edit_needs_change,
+    )
 
+    if request.method == "POST":  # POST
         _document_lock_clear(doc.id, user.username)
 
-        if which_form == "doc":
-            if doc.allows(user, "edit"):
-                post_data = request.POST.copy()
-                post_data.update({"locale": request.LANGUAGE_CODE})
+        post_data = request.POST.copy()
+        post_data.update({"locale": request.LANGUAGE_CODE})
 
-                topics = []
-                for t in post_data.getlist("topics"):
-                    topics.append(int(t))
-                post_data.setlist("topics", topics)
+        topics = []
+        for t in post_data.getlist("topics"):
+            topics.append(int(t))
+        post_data.setlist("topics", topics)
 
-                doc_form = DocumentForm(
-                    post_data,
-                    instance=doc,
-                    can_archive=can_archive,
-                    can_edit_needs_change=can_edit_needs_change,
+        doc_form = DocumentForm(
+            post_data,
+            instance=doc,
+            can_archive=can_archive,
+            can_edit_needs_change=can_edit_needs_change,
+        )
+        if doc_form.is_valid():
+            # Get the possibly new slug for the imminent redirection:
+            try:
+                doc = doc_form.save(None)
```

Comment:
Why don't we pass the logic from here to the form save method?

Reply:
There's also validation logic in the form that may be a better fit for this

Reply:
I moved the try/except/else out only to replace it with a try/except/else. I'm not 100% sure this is what you intended the outcome to be.  I looked at moving some of the other validation into the view, but it seems like there is a series of events that has to happen before each item in the save process happens. You need a real `doc` to get a `doc.pk`, you need the `save_m2m()` to get the products relationship....not sure how to make it better off hand?  Let me know your thoughts.

Reply:
I was thinking to replace everything here with just `doc = doc_form.save()` or doc = doc_form.save(commit=False)` and do all the checks in the form. The errors can live in the clean method and the object manipulation in the save. Let's remove these changes and just merge it

Reply:
Will do!